sources/api/netdog/src/networkd_status.rs

```diff
@@ -0,0 +1,304 @@
+impl NetworkdStatus {
+    pub(crate) fn primary_address(&self) -> Result<IpAddr> {
+        use error::NoIpAddressSnafu;
+        match self.addresses.len().cmp(&1) {
+            Ordering::Less => NoIpAddressSnafu {
+                interface: self.name.clone(),
+            }
+            .fail(),
+            Ordering::Equal => Ok(self.addresses[0]),
+            Ordering::Greater => {
+                for addr in self.addresses.iter() {
+                    if addr.is_ipv4() {
+                        return Ok(*addr);
+                    }
+                }
+                NoIpAddressSnafu {
+                    interface: self.name.clone(),
+                }
+                .fail()
+            }
+        }
+    }
+}
+
+#[cfg(net_backend = "systemd-networkd")]
+pub(crate) fn get_link_status(link: String) -> Result<NetworkdStatus> {
+    let systemd_networkctl_result = Command::new(NETWORKCTL)
+        .arg("status")
+        .arg("--json=pretty")
+        .arg(link)
+        .output()
+        .context(error::NetworkctlExecutionSnafu)?;
+    ensure!(
+        systemd_networkctl_result.status.success(),
+        error::FailedNetworkctlSnafu {
+            stderr: String::from_utf8_lossy(&systemd_networkctl_result.stderr)
+        }
+    );
+
+    let status_output: String = String::from_utf8(systemd_networkctl_result.stdout).unwrap();
+
+    let networkd_status = serde_json::from_str(&status_output).unwrap();
+
+    Ok(networkd_status)
+}
+
+mod error {
+    use crate::interface_id::InterfaceId;
+    use snafu::Snafu;
+    use std::io;
+
+    #[derive(Debug, Snafu)]
+    #[snafu(visibility(pub(crate)))]
+    pub(crate) enum Error {
+        #[snafu(display("Failed to run 'networkctl': {}", source))]
+        NetworkctlExecution { source: io::Error },
+
+        #[snafu(display("'networkctl' failed: {}", stderr))]
+        FailedNetworkctl { stderr: String },
+
+        #[snafu(display("Failed to parse IP Address: {:?} {}", input, msg))]
+        BadIpAddress { input: Vec<u8>, msg: String },
+
+        #[snafu(display("No IP Address for Primary Interface: {:?}", interface))]
+        NoIpAddress { interface: InterfaceId },
+    }
+}
+
+pub(crate) use error::Error;
+#[cfg(net_backend = "systemd-networkd")]
+type Result<T> = std::result::Result<T, error::Error>;
+
+#[cfg(test)]
+#[cfg(net_backend = "systemd-networkd")]
+mod tests {
+    use super::*;
+    use std::fs;
+    use std::net::Ipv4Addr;
+    use std::path::Path;
+    use std::path::PathBuf;
+
+    fn test_data() -> PathBuf {
+        PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("test_data")
+    }
+
+    fn networkd_config() -> PathBuf {
+        test_data().join("networkd")
+    }
+
+    fn read_output_file<P>(path: P) -> String
+    where
+        P: AsRef<Path>,
+    {
+        let path = path.as_ref();
+        let path_str = fs::read_to_string(path).unwrap();
+
+        path_str
+    }
+
+    // full deserialize test
+    #[test]
+    fn parse_qemu_status() {
```

Comment:
I'm a little confused about this: is the output different?  It's a bit odd to have a test that references qemu directly.